File: ShapeHandler/src/shapehandler/code/java/3DShapes.java
package shapehandler.code.java;

import static java.lang.Math.PI;
import static java.lang.Math.sqrt;

class Cube {

    private double side;

    public Cube(double side) {
        this.side = side;
    }

    public double volume() {
        return Math.pow(side, 3);
    }

    public double spaceDiagonal() {
        return sqrt(3) * side;
    }

    public double surfaceArea() { return 6 * Math.pow(side, 2); }
}

class Cuboid {

    private double length, width, height;

    public Cuboid (double length, double width, double height) {
        this.length = length;
        this.width = width;
        this.height = height;
    }

    public double volume() {
        return length * width * height;
    }

    public double spaceDiagonal() {
        return sqrt(Math.pow(length, 2) + Math.pow(width, 2) + Math.pow(height, 2));
    }

    public double surfaceArea() { return 2 * ((width * length) + (height * length) + (height * width)); }
}

class Cylinder {

    private double radius, height;

    public Cylinder(double radius, double height) {
        this.radius = radius;
        this.height = height;
    }

    public double volume() {
        return PI * Math.pow(radius, 2) * height;
    }

    public double surfaceArea() {
        return (2 * PI * radius * height) + (2 * PI * Math.pow(radius, 2));
    }

    public double lateralSurfaceArea() {
        return 2 * PI * radius * height;
    }

    public double baseArea() {
        return PI * Math.pow(radius, 2);
    }
}

class Pyramid {

    private double length, width, height;

    public Pyramid(double length, double width, double height) {
        this.length = length;
        this.width = width;
        this.height = height;
    }

    public double volume() {
        return (length * width * height) / 3;
    }

    public double surfaceArea() {
        return ((length * width) + length * sqrt(Math.pow(width/2, 2) + Math.pow(height, 2))) + (width * sqrt(Math.pow(length/2, 2) + Math.pow(height, 2)));
    }

    public double lateralSurfaceArea() {
        return (length * sqrt(Math.pow(width/2, 2) + Math.pow(height, 2))) + (width * sqrt(Math.pow(length/2, 2) + Math.pow(height, 2)));
    }

    public double baseArea() {
        return length * width;
    }
}

class Sphere {

    private double radius;

    public Sphere(double radius) {
        this.radius = radius;
    }

    public double volume() {
        return 1.333333 * PI * Math.pow(radius, 3); // rough estimate
    }

    public double diameter() {
        return 2 * radius;
    }

    public double surfaceArea() {
        return 4 * PI * Math.pow(radius, 2);
    }
}

File: ShapeHandler/src/shapehandler/code/kotlin/3DShapes.kt
package shapehandler.code.kotlin

import kotlin.math.PI
import kotlin.math.pow
import kotlin.math.sqrt

class Cube(var side: Double) {
    fun volume(): Double {
        return side.pow(3)
    }

    fun spaceDiagonal(): Double {
        return sqrt(3.0) * side
    }

    fun surfaceArea(): Double {
        return 6 * side.pow(2)
    }
}

class Cuboid(var length: Double, var width: Double, var height: Double) {
    fun volume(): Double {
        return length * width * height
    }

    fun spaceDiagonal(): Double {
        return sqrt(length.pow(2.0) + width.pow(2.0) + height.pow(2.0))
    }

    fun surfaceArea(): Double {
        return 2 * (length * width + width * height + length * height)
    }
}

class Cylinder(var radius: Double, var height: Double) {
    fun volume(): Double {
        return PI * radius.pow(2) * height
    }

    fun surfaceArea(): Double {
        return (2 * PI * radius * height) + (2 * PI * radius.pow(2))
    }

    fun lateralSurfaceArea(): Double {
        return 2 * PI * radius * height
    }

    fun baseArea(): Double {
        return PI * radius.pow(2);
    }
}

class Pyramid(var length: Double, var width: Double, var height: Double) {
    fun volume(): Double {
        return (length * width * height) / 3
    }

    fun surfaceArea(): Double {
        return length * width + length * sqrt((width / 2).pow(2.0) + height.pow(2.0)) + width * sqrt(
            (length / 2).pow(
                2.0
            ) + height.pow(2.0)
        )
    }

    fun lateralSurfaceArea(): Double {
        return (length * sqrt((width / 2).pow(2.0) + height.pow(2.0))) +
                (width * sqrt((length / 2).pow(2.0) + height.pow(2.0)))
    }

    fun baseArea(): Double {
        return length * width
    }
}

class Sphere(var radius: Double) {
    fun volume(): Double {
        return 1.333333 * PI * radius.pow(3) // rough estimate
    }

    fun diameter(): Double {
        return 2 * radius
    }

    fun surfaceArea(): Double {
        return 4 * PI * radius.pow(2)
    }
}